// @bitgo/utxo-lib/dist/src/noble_ecc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.musig = exports.bip32 = exports.ECPair = exports.ecc = void 0;
const createHash = require("create-hash");
const createHmac = require("create-hmac");
const ecpair_1 = require("ecpair");
const necc = require("@noble/secp256k1");
const bip32_1 = require("bip32");
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore base_crypto is exported as a subPath export, ignoring since compiler complains about importing like this
const baseCrypto = require("@brandonblack/musig/base_crypto");
const musig_1 = require("@brandonblack/musig");
necc.utils.sha256Sync = (...messages) => {
    const sha256 = createHash('sha256');
    for (const message of messages)
        sha256.update(message);
    return sha256.digest();
};
necc.utils.hmacSha256Sync = (key, ...messages) => {
    const hash = createHmac('sha256', Buffer.from(key));
    messages.forEach((m) => hash.update(m));
    return Uint8Array.from(hash.digest());
};
const defaultTrue = (param) => param !== false;
function throwToNull(fn) {
    try {
        return fn();
    }
    catch (e) {
        return null;
    }
}
function isPoint(p, xOnly) {
    if ((p.length === 32) !== xOnly)
        return false;
    try {
        return !!necc.Point.fromHex(p);
    }
    catch (e) {
        return false;
    }
}
function toBigInt(b) {
    const buff = Buffer.from(b);
    if (buff.length !== 32) {
        throw new Error('Invalid size ${buff.length}');
    }
    return BigInt(`0x${buff.toString('hex')}`);
}
const ecc = {
    isPoint: (p) => isPoint(p, false),
    isPrivate: (d) => necc.utils.isValidPrivateKey(d),
    isXOnlyPoint: (p) => isPoint(p, true),
    xOnlyPointAddTweak: (p, tweak) => throwToNull(() => {
        const P = necc.utils.pointAddScalar(p, tweak, true);
        const parity = P[0] % 2 === 1 ? 1 : 0;
        return { parity, xOnlyPubkey: P.slice(1) };
    }),
    pointFromScalar: (sk, compressed) => throwToNull(() => necc.getPublicKey(sk, defaultTrue(compressed))),
    pointCompress: (p, compressed) => {
        return necc.Point.fromHex(p).toRawBytes(defaultTrue(compressed));
    },
    pointMultiply: (a, tweak, compressed) => throwToNull(() => necc.utils.pointMultiply(a, tweak, defaultTrue(compressed))),
    pointAdd: (a, b, compressed) => throwToNull(() => {
        const A = necc.Point.fromHex(a);
        const B = necc.Point.fromHex(b);
        return A.add(B).toRawBytes(defaultTrue(compressed));
    }),
    pointAddScalar: (p, tweak, compressed) => throwToNull(() => necc.utils.pointAddScalar(p, tweak, defaultTrue(compressed))),
    privateAdd: (d, tweak) => throwToNull(() => {
        const res = necc.utils.privateAdd(d, tweak);
        // tiny-secp256k1 returns null rather than allowing a 0 private key to be returned
        // ECPair.testEcc() requires that behavior.
        if (res === null || res === void 0 ? void 0 : res.every((i) => i === 0))
            return null;
        return res;
    }),
    privateNegate: (d) => necc.utils.privateNegate(d),
    sign: (h, d, e) => {
        return necc.signSync(h, d, { der: false, extraEntropy: e });
    },
    signSchnorr: (h, d, e = Buffer.alloc(32, 0x00)) => {
        return necc.schnorr.signSync(h, d, e);
    },
    verify: (h, Q, signature, strict) => {
        return necc.verify(signature, h, Q, { strict });
    },
    verifySchnorr: (h, Q, signature) => {
        return necc.schnorr.verifySync(signature, h, Q);
    },
};
exports.ecc = ecc;
const crypto = {
    ...baseCrypto,
    pointMultiplyUnsafe(p, a, compress) {
        try {
            const product = necc.Point.fromHex(p).multiplyAndAddUnsafe(necc.Point.ZERO, toBigInt(a), BigInt(1));
            if (!product)
                return null;
            return product.toRawBytes(compress);
        }
        catch {
            return null;
        }
    },
    pointMultiplyAndAddUnsafe(p1, a, p2, compress) {
        try {
            const p2p = necc.Point.fromHex(p2);
            const p = necc.Point.fromHex(p1).multiplyAndAddUnsafe(p2p, toBigInt(a), BigInt(1));
            if (!p)
                return null;
            return p.toRawBytes(compress);
        }
        catch {
            return null;
        }
    },
    pointAdd(a, b, compress) {
        try {
            return necc.Point.fromHex(a).add(necc.Point.fromHex(b)).toRawBytes(compress);
        }
        catch {
            return null;
        }
    },
    pointAddTweak(p, tweak, compress) {
        try {
            const P = necc.Point.fromHex(p);
            const t = baseCrypto.readSecret(tweak);
            const Q = necc.Point.BASE.multiplyAndAddUnsafe(P, t, BigInt(1));
            if (!Q)
                throw new Error('Tweaked point at infinity');
            return Q.toRawBytes(compress);
        }
        catch {
            return null;
        }
    },
    pointCompress(p, compress = true) {
        return necc.Point.fromHex(p).toRawBytes(compress);
    },
    liftX(p) {
        try {
            return necc.Point.fromHex(p).toRawBytes(false);
        }
        catch {
            return null;
        }
    },
    getPublicKey(s, compress) {
        try {
            return necc.getPublicKey(s, compress);
        }
        catch {
            return null;
        }
    },
    taggedHash: necc.utils.taggedHashSync,
    sha256(...messages) {
        const sha256 = createHash('sha256');
        for (const message of messages)
            sha256.update(message);
        return sha256.digest();
    },
};
const ECPair = ecpair_1.ECPairFactory(ecc);
exports.ECPair = ECPair;
const bip32 = bip32_1.BIP32Factory(ecc);
exports.bip32 = bip32;
const musig = musig_1.MuSigFactory(crypto);
exports.musig = musig;
