// @bitgo/utxo-lib/dist/src/bitgo/wallet/WalletKeys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RootWalletKeys = exports.DerivedWalletKeys = exports.WalletKeys = exports.eqPublicKey = void 0;
function eqPublicKey(a, b) {
    return a.publicKey.equals(b.publicKey);
}
exports.eqPublicKey = eqPublicKey;
/**
 * Base class for RootWalletKeys and DerivedWalletKeys.
 * Keys can be either public keys or private keys.
 */
class WalletKeys {
    /**
     * @param triple - bip32 key triple
     */
    constructor(triple) {
        this.triple = triple;
        triple.forEach((a, i) => {
            triple.forEach((b, j) => {
                if (eqPublicKey(a, b) && i !== j) {
                    throw new Error(`wallet keys must be distinct`);
                }
            });
        });
        this.publicKeys = this.triple.map((k) => k.publicKey);
    }
    get user() {
        return this.triple[0];
    }
    get backup() {
        return this.triple[1];
    }
    get bitgo() {
        return this.triple[2];
    }
}
exports.WalletKeys = WalletKeys;
/**
 * Set of WalletKeys derived from RootWalletKeys. Suitable for signing transaction inputs.
 * Contains reference to the RootWalletKeys this was derived from as well as the paths used
 * for derivation.
 */
class DerivedWalletKeys extends WalletKeys {
    /**
     * @param parent - wallet keys to derive from
     * @param paths - paths to derive with
     */
    constructor(parent, paths) {
        super(parent.triple.map((k, i) => k.derivePath(paths[i])));
        this.parent = parent;
        this.paths = paths;
    }
}
exports.DerivedWalletKeys = DerivedWalletKeys;
/**
 * Set of root wallet keys, typically instantiated using the wallet xpub triple.
 */
class RootWalletKeys extends WalletKeys {
    /**
     * @param triple - bip32 key triple
     * @param derivationPrefixes - Certain v1 wallets or their migrated v2 counterparts
     *                             can have a nonstandard prefix.
     */
    constructor(triple, derivationPrefixes = [
        RootWalletKeys.defaultPrefix,
        RootWalletKeys.defaultPrefix,
        RootWalletKeys.defaultPrefix,
    ]) {
        super(triple);
        this.derivationPrefixes = derivationPrefixes;
        derivationPrefixes.forEach((p) => {
            if (p.startsWith('/') || p.endsWith('/')) {
                throw new Error(`derivation prefix must not start or end with a slash`);
            }
        });
    }
    /**
     * @param key
     * @param chain
     * @param index
     * @return full derivation path for key, including key-specific prefix
     */
    getDerivationPath(key, chain, index) {
        if (!this.derivationPrefixes) {
            throw new Error(`no derivation prefixes`);
        }
        const prefix = this.derivationPrefixes.find((prefix, i) => eqPublicKey(key, this.triple[i]));
        if (prefix === undefined) {
            throw new Error(`key not in walletKeys`);
        }
        return `${prefix}/${chain}/${index}`;
    }
    /**
     * @param chain
     * @param index
     * @return walletKeys for a particular address identified by (chain, index)
     */
    deriveForChainAndIndex(chain, index) {
        return new DerivedWalletKeys(this, this.triple.map((k) => this.getDerivationPath(k, chain, index)));
    }
}
exports.RootWalletKeys = RootWalletKeys;
RootWalletKeys.defaultPrefix = '0/0';
