// @bitgo/utxo-lib/dist/src/bitgo/wallet/WalletUnspentSigner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletUnspentSigner = void 0;
const WalletKeys_1 = require("./WalletKeys");
class WalletUnspentSigner {
    constructor(walletKeys, signer, cosigner) {
        this.signer = signer;
        this.cosigner = cosigner;
        if (Array.isArray(walletKeys)) {
            walletKeys = new WalletKeys_1.RootWalletKeys(walletKeys);
        }
        this.signerIndex = walletKeys.triple.findIndex((k) => WalletKeys_1.eqPublicKey(k, signer));
        if (this.signerIndex === undefined) {
            throw new Error(`signer not part of walletKeys`);
        }
        this.cosignerIndex = walletKeys.triple.findIndex((k) => WalletKeys_1.eqPublicKey(k, cosigner));
        if (this.cosignerIndex === undefined) {
            throw new Error(`cosigner not part of walletKeys`);
        }
        this.walletKeys = walletKeys;
        if (WalletKeys_1.eqPublicKey(signer, cosigner)) {
            throw new Error(`signer must not equal cosigner`);
        }
        if (signer.isNeutered()) {
            throw new Error(`signer must have private key`);
        }
    }
    static from(walletKeys, signer, cosigner) {
        return new WalletUnspentSigner(walletKeys, signer, cosigner);
    }
    /**
     * @param chain
     * @param index
     * @return WalletUnspentSigner that contains keys for generating output scripts and signatures.
     */
    deriveForChainAndIndex(chain, index) {
        if (this.walletKeys instanceof WalletKeys_1.DerivedWalletKeys) {
            throw new Error(`cannot derive again from DerivedWalletKeys`);
        }
        if (this.walletKeys instanceof WalletKeys_1.RootWalletKeys) {
            return new WalletUnspentSigner(this.walletKeys.deriveForChainAndIndex(chain, index), this.signer.derivePath(this.walletKeys.getDerivationPath(this.signer, chain, index)), this.cosigner.derivePath(this.walletKeys.getDerivationPath(this.cosigner, chain, index)));
        }
        throw new Error(`invalid state`);
    }
}
exports.WalletUnspentSigner = WalletUnspentSigner;
