// chronik-client/dist/test/fixtures.js
"use strict";
// @generated
Object.defineProperty(exports, "__esModule", { value: true });
exports.VALIDATION_UTXOS = exports.TEST_TX_IN_NODE = exports.TEST_TX_NNG = exports.GENESIS_UTXO_IN_NODE = exports.GENESIS_UTXO = exports.GENESIS_TX_IN_NODE = exports.GENESIS_TX = exports.GENESIS_BLOCK_DETAILS = exports.GENESIS_BLOCK_INFO = exports.GENESIS_PK = void 0;
exports.GENESIS_PK = '04678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc' +
    '3f4cef38c4f35504e51ec112de5c384df7ba0b8d578a4c702b6bf11d5f';
exports.GENESIS_BLOCK_INFO = {
    hash: '000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f',
    prevHash: '0000000000000000000000000000000000000000000000000000000000000000',
    height: 0,
    nBits: 0x1d00ffff,
    timestamp: '1231006505',
    blockSize: '285',
    numTxs: '1',
    numInputs: '1',
    numOutputs: '1',
    sumInputSats: '0',
    sumCoinbaseOutputSats: '5000000000',
    sumNormalOutputSats: '0',
    sumBurnedSats: '0',
};
exports.GENESIS_BLOCK_DETAILS = {
    version: 1,
    merkleRoot: '4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b',
    nonce: '2083236893',
    medianTimestamp: '1231006505',
};
exports.GENESIS_TX = {
    txid: '4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b',
    version: 1,
    inputs: [
        {
            prevOut: {
                txid: '0000000000000000000000000000000000000000000000000000000000000000',
                outIdx: 0xffffffff,
            },
            inputScript: '04ffff001d0104455468652054696d65732030332f4a616e2f32303039204368616e' +
                '63656c6c6f72206f6e206272696e6b206f66207365636f6e64206261696c6f757420' +
                '666f722062616e6b73',
            outputScript: undefined,
            value: '0',
            sequenceNo: 0xffffffff,
            slpBurn: undefined,
            slpToken: undefined,
        },
    ],
    outputs: [
        {
            value: '5000000000',
            outputScript: '41' + exports.GENESIS_PK + 'ac',
            slpToken: undefined,
            spentBy: undefined,
        },
    ],
    lockTime: 0,
    slpTxData: undefined,
    slpErrorMsg: undefined,
    block: {
        hash: '000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f',
        height: 0,
        timestamp: '1231006505',
    },
    size: 204,
    isCoinbase: true,
    timeFirstSeen: '0',
    network: 'XEC',
};
exports.GENESIS_TX_IN_NODE = {
    txid: '4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b',
    version: 1,
    inputs: [
        {
            prevOut: {
                txid: '0000000000000000000000000000000000000000000000000000000000000000',
                outIdx: 0xffffffff,
            },
            inputScript: '04ffff001d0104455468652054696d65732030332f4a616e2f32303039204368616e' +
                '63656c6c6f72206f6e206272696e6b206f66207365636f6e64206261696c6f757420' +
                '666f722062616e6b73',
            outputScript: undefined,
            value: '0',
            sequenceNo: 0xffffffff,
            slpBurn: undefined,
            slpToken: undefined,
        },
    ],
    outputs: [
        {
            value: '5000000000',
            outputScript: '41' + exports.GENESIS_PK + 'ac',
            slpToken: undefined,
            spentBy: undefined,
        },
    ],
    lockTime: 0,
    slpTxData: undefined,
    slpErrorMsg: undefined,
    block: {
        hash: '000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f',
        height: 0,
        timestamp: '1231006505',
    },
    size: 204,
    isCoinbase: true,
    timeFirstSeen: '0',
};
exports.GENESIS_UTXO = {
    outpoint: { txid: exports.GENESIS_TX.txid, outIdx: 0 },
    blockHeight: 0,
    isCoinbase: true,
    value: '5000000000',
    slpMeta: undefined,
    slpToken: undefined,
    network: 'XEC',
};
exports.GENESIS_UTXO_IN_NODE = {
    outpoint: { txid: exports.GENESIS_TX.txid, outIdx: 0 },
    blockHeight: 0,
    isCoinbase: true,
    value: '5000000000',
    slpMeta: undefined,
    slpToken: undefined,
    isFinal: true,
};
exports.TEST_TX_NNG = {
    txid: '0f3c3908a2ddec8dea91d2fe1f77295bbbb158af869bff345d44ae800f0a5498',
    version: 2,
    inputs: [
        {
            inputScript: '473044022042349e5e9e58c4c7b1fc9cbcdd1a1c9774b2ae95b7704ce40c7058ef' +
                '14ba2854022022ba0cd4ead86982e7fc090ad06569a72c165efa3634feac4e722e' +
                'ee405b92674121022d577b731fb05971d54951e4cb8bd11120327eba3e0fdd5a4d' +
                '18f74a882df1a4',
            outputScript: '76a914d15b9793d6af77663f8acf7e2c884f114ef901da88ac',
            prevOut: {
                txid: 'fbe9a326ea525013e11e3f6341b3e03d6ad189bac30f94d5ceaf24ef4388e069',
                outIdx: 1,
            },
            sequenceNo: 4294967295,
            slpBurn: undefined,
            slpToken: {
                amount: '50000000',
                isMintBaton: false,
            },
            value: '546',
        },
        {
            inputScript: '473044022076e0c764b7d5a5f738304fec1df97db4ba6fc35f8ccb2438de8aabff' +
                '781edb5c022002016344d4432a5837569b9c764b88aa71f98dbfb76c48cbd0ad9f' +
                '18dc15be0e4121022d577b731fb05971d54951e4cb8bd11120327eba3e0fdd5a4d' +
                '18f74a882df1a4',
            outputScript: '76a914d15b9793d6af77663f8acf7e2c884f114ef901da88ac',
            prevOut: {
                txid: '7ded3b52bbe20fc40d147a4e799fda91e589d29e017789ab6361e462dc972094',
                outIdx: 0,
            },
            sequenceNo: 4294967295,
            slpBurn: undefined,
            slpToken: undefined,
            value: '100000',
        },
    ],
    outputs: [
        {
            outputScript: '6a04534c500001010453454e44200daf200e3418f2df1158efef36fbb507f12928' +
                'f1fdcf3543703e64e75a4a90730800000000004c4b40080000000002aea540',
            slpToken: undefined,
            spentBy: undefined,
            value: '0',
        },
        {
            outputScript: '76a9149c371def7e7cf89b30a62d658147937e679a965388ac',
            slpToken: {
                amount: '5000000',
                isMintBaton: false,
            },
            spentBy: {
                txid: '962ace9db1a36d06c129dffbe9a92bcf2eafe37d1a44aedfc6f957d2be69f149',
                outIdx: 0,
            },
            value: '546',
        },
        {
            outputScript: '76a914e7b4f63ec550ada1aed74960ddc4e0e107cd6cd188ac',
            slpToken: {
                amount: '45000000',
                isMintBaton: false,
            },
            spentBy: {
                txid: '11ce5e4249c5b43927810129d887ce0df3bbde46d036998dff0180f94d2df6f8',
                outIdx: 0,
            },
            value: '546',
        },
        {
            outputScript: '76a914d15b9793d6af77663f8acf7e2c884f114ef901da88ac',
            slpToken: undefined,
            spentBy: {
                txid: '11ce5e4249c5b43927810129d887ce0df3bbde46d036998dff0180f94d2df6f8',
                outIdx: 1,
            },
            value: '98938',
        },
    ],
    lockTime: 0,
    slpTxData: {
        genesisInfo: undefined,
        slpMeta: {
            tokenId: '0daf200e3418f2df1158efef36fbb507f12928f1fdcf3543703e64e75a4a9073',
            tokenType: 'FUNGIBLE',
            txType: 'SEND',
            groupTokenId: undefined,
        },
    },
    slpErrorMsg: undefined,
    block: {
        hash: '0000000000000000452f19532a6297ea194eaacac6d3bbcbf7c08a74cad84b44',
        height: 697728,
        timestamp: '1627790415',
    },
    size: 479,
    isCoinbase: false,
    timeFirstSeen: '0',
    network: 'XEC',
};
exports.TEST_TX_IN_NODE = {
    txid: '0f3c3908a2ddec8dea91d2fe1f77295bbbb158af869bff345d44ae800f0a5498',
    version: 2,
    inputs: [
        {
            inputScript: '473044022042349e5e9e58c4c7b1fc9cbcdd1a1c9774b2ae95b7704ce40c7058ef' +
                '14ba2854022022ba0cd4ead86982e7fc090ad06569a72c165efa3634feac4e722e' +
                'ee405b92674121022d577b731fb05971d54951e4cb8bd11120327eba3e0fdd5a4d' +
                '18f74a882df1a4',
            outputScript: '76a914d15b9793d6af77663f8acf7e2c884f114ef901da88ac',
            prevOut: {
                txid: 'fbe9a326ea525013e11e3f6341b3e03d6ad189bac30f94d5ceaf24ef4388e069',
                outIdx: 1,
            },
            sequenceNo: 4294967295,
            slpBurn: undefined,
            slpToken: undefined,
            value: '546',
        },
        {
            inputScript: '473044022076e0c764b7d5a5f738304fec1df97db4ba6fc35f8ccb2438de8aabff' +
                '781edb5c022002016344d4432a5837569b9c764b88aa71f98dbfb76c48cbd0ad9f' +
                '18dc15be0e4121022d577b731fb05971d54951e4cb8bd11120327eba3e0fdd5a4d' +
                '18f74a882df1a4',
            outputScript: '76a914d15b9793d6af77663f8acf7e2c884f114ef901da88ac',
            prevOut: {
                txid: '7ded3b52bbe20fc40d147a4e799fda91e589d29e017789ab6361e462dc972094',
                outIdx: 0,
            },
            sequenceNo: 4294967295,
            slpBurn: undefined,
            slpToken: undefined,
            value: '100000',
        },
    ],
    outputs: [
        {
            outputScript: '6a04534c500001010453454e44200daf200e3418f2df1158efef36fbb507f12928' +
                'f1fdcf3543703e64e75a4a90730800000000004c4b40080000000002aea540',
            slpToken: undefined,
            spentBy: undefined,
            value: '0',
        },
        {
            outputScript: '76a9149c371def7e7cf89b30a62d658147937e679a965388ac',
            slpToken: undefined,
            spentBy: {
                txid: '962ace9db1a36d06c129dffbe9a92bcf2eafe37d1a44aedfc6f957d2be69f149',
                outIdx: 0,
            },
            value: '546',
        },
        {
            outputScript: '76a914e7b4f63ec550ada1aed74960ddc4e0e107cd6cd188ac',
            slpToken: undefined,
            spentBy: {
                txid: '11ce5e4249c5b43927810129d887ce0df3bbde46d036998dff0180f94d2df6f8',
                outIdx: 0,
            },
            value: '546',
        },
        {
            outputScript: '76a914d15b9793d6af77663f8acf7e2c884f114ef901da88ac',
            slpToken: undefined,
            spentBy: {
                txid: '11ce5e4249c5b43927810129d887ce0df3bbde46d036998dff0180f94d2df6f8',
                outIdx: 1,
            },
            value: '98938',
        },
    ],
    lockTime: 0,
    slpTxData: undefined,
    slpErrorMsg: undefined,
    block: {
        hash: '0000000000000000452f19532a6297ea194eaacac6d3bbcbf7c08a74cad84b44',
        height: 697728,
        timestamp: '1627790415',
    },
    size: 479,
    isCoinbase: false,
    timeFirstSeen: '0',
};
exports.VALIDATION_UTXOS = [
    {
        txid: '4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b',
        outIdx: 0,
    },
    {
        txid: '0f3c3908a2ddec8dea91d2fe1f77295bbbb158af869bff345d44ae800f0a5498',
        outIdx: 1,
    },
    {
        txid: '0f3c3908a2ddec8dea91d2fe1f77295bbbb158af869bff345d44ae800f0a5498',
        outIdx: 100,
    },
    {
        txid: '0000000000000000000000000000000000000000000000000000000000000000',
        outIdx: 100,
    },
];
