// @bitgo/utxo-lib/dist/src/bitgo/bitcoincash/address.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toOutputScriptWithFormat = exports.fromOutputScriptWithFormat = exports.toOutputScriptFromCashAddress = exports.fromOutputScriptToCashAddress = exports.getPrefix = void 0;
/**
 * Wrapper around `cashaddress` library.
 *
 * Performs some address sanitation:
 * - add prefix if missing
 * - normalize to lower-case
 * - reject mixed-case
 *
 * Based on these documents
 *
 * - https://github.com/bitcoincashorg/bitcoincash.org/blob/master/spec/cashaddr.md
 * - https://www.bitcoinabc.org/cashaddr/
 */
const cashaddress = require("cashaddress");
const bitcoinjs = require("bitcoinjs-lib");
const networks_1 = require("../../networks");
/**
 * @param name
 * @param output
 * @return the encoded pubkeyhash or scripthash
 */
function getHashFromOutputScript(name, output) {
    const func = bitcoinjs.payments[name];
    if (!func) {
        throw new Error(`no payment with name ${name}`);
    }
    try {
        return func({ output }).hash;
    }
    catch (e) {
        return undefined;
    }
}
/**
 * @param network
 * @return network-specific cashaddr prefix
 */
function getPrefix(network) {
    switch (network) {
        case networks_1.networks.bitcoincash:
            return 'bitcoincash';
        case networks_1.networks.bitcoincashTestnet:
            return 'bchtest';
        case networks_1.networks.ecash:
            return 'ecash';
        case networks_1.networks.ecashTest:
            return 'ectest';
        default:
            throw new Error(`unsupported prefix for ${networks_1.getNetworkName(network)}`);
    }
}
exports.getPrefix = getPrefix;
/**
 * @param outputScript
 * @param network
 * @return outputScript encoded as cashaddr (prefixed, lowercase)
 */
function fromOutputScriptToCashAddress(outputScript, network) {
    if (!networks_1.isBitcoinCash(network) && !networks_1.isECash(network)) {
        throw new Error(`invalid network`);
    }
    for (const [paymentName, scriptType] of [
        ['p2pkh', 'pubkeyhash'],
        ['p2sh', 'scripthash'],
    ]) {
        const hash = getHashFromOutputScript(paymentName, outputScript);
        if (hash) {
            return cashaddress.encode(getPrefix(network), scriptType, hash);
        }
    }
    throw new Error(`could not determine hash for outputScript`);
}
exports.fromOutputScriptToCashAddress = fromOutputScriptToCashAddress;
/**
 * @param address - Accepts addresses with and without prefix. Accepts all-lowercase and all-uppercase addresses. Rejects mixed-case addresses.
 * @param network
 * @return decoded output script
 */
function toOutputScriptFromCashAddress(address, network) {
    if (!networks_1.isBitcoinCash(network) && !networks_1.isECash(network)) {
        throw new Error(`invalid network`);
    }
    if (address === address.toUpperCase()) {
        address = address.toLowerCase();
    }
    if (address !== address.toLowerCase()) {
        throw new Error(`mixed-case addresses not allowed`);
    }
    if (!address.startsWith(getPrefix(network) + ':')) {
        address = `${getPrefix(network)}:${address}`;
    }
    const decoded = cashaddress.decode(address);
    let outputScript;
    switch (decoded.version) {
        case 'scripthash':
            outputScript = bitcoinjs.payments.p2sh({ hash: decoded.hash }).output;
            break;
        case 'pubkeyhash':
            outputScript = bitcoinjs.payments.p2pkh({ hash: decoded.hash }).output;
            break;
        default:
            throw new Error(`unknown version ${decoded.version}`);
    }
    if (!outputScript) {
        throw new Error(`could not determine output script`);
    }
    return outputScript;
}
exports.toOutputScriptFromCashAddress = toOutputScriptFromCashAddress;
/**
 * @param outputScript
 * @param format
 * @param network
 * @return address in specified format
 */
function fromOutputScriptWithFormat(outputScript, format, network) {
    if (!networks_1.isBitcoinCash(network) && !networks_1.isECash(network)) {
        throw new Error(`invalid network`);
    }
    if (format === 'cashaddr') {
        return fromOutputScriptToCashAddress(outputScript, network);
    }
    if (format === 'default') {
        return bitcoinjs.address.fromOutputScript(outputScript, network);
    }
    throw new Error(`invalid format`);
}
exports.fromOutputScriptWithFormat = fromOutputScriptWithFormat;
/**
 * @param address
 * @param format
 * @param network
 * @return output script from address in specified format
 */
function toOutputScriptWithFormat(address, format, network) {
    if (!networks_1.isBitcoinCash(network) && !networks_1.isECash(network)) {
        throw new Error(`invalid network`);
    }
    if (format === 'cashaddr') {
        return toOutputScriptFromCashAddress(address, network);
    }
    if (format === 'default') {
        return bitcoinjs.address.toOutputScript(address, network);
    }
    throw new Error(`invalid format`);
}
exports.toOutputScriptWithFormat = toOutputScriptWithFormat;
